# Deep Learning 설계하기

In [ ]:
# !pip install tensorflow

   ---------------------------------------- 0.0/332.0 MB ? eta -:--:--
    --------------------------------------- 5.0/332.0 MB 23.5 MB/s eta 0:00:14
   --- ------------------------------------ 27.3/332.0 MB 66.4 MB/s eta 0:00:05
   --- ------------------------------------ 30.1/332.0 MB 48.0 MB/s eta 0:00:07
   ----- ---------------------------------- 44.8/332.0 MB 54.0 MB/s eta 0:00:06
   -------- ------------------------------- 66.8/332.0 MB 64.6 MB/s eta 0:00:05
   ---------- ----------------------------- 84.9/332.0 MB 71.2 MB/s eta 0:00:04
   ----------- ---------------------------- 99.1/332.0 MB 68.8 MB/s eta 0:00:04
   -------------- ------------------------ 120.6/332.0 MB 73.8 MB/s eta 0:00:03
   --------------- ----------------------- 130.0/332.0 MB 74.6 MB/s eta 0:00:03
   --------------- ----------------------- 132.4/332.0 MB 64.6 MB/s eta 0:00:04
   ----------------- --------------------- 149.9/332.0 MB 68.2 MB/s eta 0:00:03
   ------------------ -------------------- 156.8/3

## 환경 준비

In [5]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy as np

c:\ProgramData\anaconda3\Lib\site-packages\keras\src\export\tf2onnx_lib.py:8: FutureWarning: In the future `np.object` will be defined as the corresponding NumPy scalar.
  if not hasattr(np, "object"):


## 데이터 준비
- Diagnosis	1–3 (범주형)	진단 코드 (암/질환 타입 분류)
- FVC	실수값	Forced Vital Capacity — 노력성 폐활량
- FEV1	실수값	Forced Expiratory Volume — 1초간 호기량
- Performance Status	0–2	WHO 수행능력지수 (일상생활 가능 정도)
- Pain	T/F	흉통 여부
- Hemoptysis	T/F	객혈(피 섞인 가래) 여부
- Dyspnoea	T/F	호흡곤란 여부
- Cough	T/F	기침
- Weakness	T/F	허약감
- T-Stage	1–4	종양 크기·침윤 정도 (TNM stage)
- N-Stage	0–3	림프절 전이 (TNM stage)
- Diabetes Mellitus	T/F	당뇨병 여부
- MI Up To 6 Months	T/F	최근 6개월 내 심근경색 여부
- PAD	T/F	말초동맥질환 여부
- Smoking	T/F	흡연 여부
- Asthma	T/F	천식 여부
- Risk1Y (Label)	T/F	수술 후 1년 내 사망 → True, 생존 → False

In [33]:
ds = np.loadtxt("ThoraricSurgery3.csv", delimiter=',')
x = ds[:, 0:16]
y = ds[:, 16]
print(type(ds), ds.shape)
print(type(x), x.shape)
print(type(y), y.shape)

<class 'numpy.ndarray'> (470, 17)
<class 'numpy.ndarray'> (470, 16)
<class 'numpy.ndarray'> (470,)


## 구조 결정

In [20]:
model = Sequential()
model.add(Dense(30, input_dim=16, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

## 모델 실행

In [22]:
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
history = model.fit(x, y, epochs=1, batch_size=16)

30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8511 - loss: 0.4047   


## 평가

In [ ]:
result_list = model.evaluate(x, y)
print(type(result_list), result_list)
print("Accuracy = %.4f" % (result_list[1]), "Loss = %.4f" % (result_list[0]))

15/15 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8532 - loss: 0.3981 
<class 'list'> [0.3981267213821411, 0.8531914949417114]
Accuracy = 0.8532
Loss = 0.3981


# TEST

In [35]:
def test_model(x, y, epochs, nodes):
    x_column_count = x.shape[1]
    model = Sequential()
    model.add(Dense(nodes, input_dim=x_column_count, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    history = model.fit(x, y, epochs=epochs, batch_size=x_column_count)
    result_list = model.evaluate(x, y)
    print("epochs = %d" % epochs, "nodes = %d" % nodes, ":","Accuracy = %.4f" % (result_list[1]), "Loss = %.4f" % (result_list[0]))

## epochs = [30, 60, 90], node = 30
- 30 : Accuracy = 0.8489 Loss = 0.3883
- 60 : Accuracy = 0.8404 Loss = 0.4088
- 90 : Accuracy = 0.8489 Loss = 0.3854

In [ ]:
epochs_list = [30, 60, 90]
for epochs in epochs_list:
    test_model(x, y, epochs, 30)

## node = [30, 60, 90], epochs = 30
- 30 : Accuracy = 0.8511 Loss = 0.3977
- 60 : Accuracy = 0.8511 Loss = 0.3909
- 90 : Accuracy = 0.8489 Loss = 0.3944

In [37]:
nodes_list = [30, 60, 90]
for nodes in nodes_list:
    test_model(x, y, 30, nodes)

Epoch 1/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7915 - loss: 0.5495   
Epoch 2/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8489 - loss: 0.4557 
Epoch 3/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8511 - loss: 0.4437 
Epoch 4/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8489 - loss: 0.4288 
Epoch 5/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8489 - loss: 0.4371 
Epoch 6/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8553 - loss: 0.4332 
Epoch 7/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8511 - loss: 0.4289 
Epoch 8/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8511 - loss: 0.4205 
Epoch 9/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8447 - loss: 0.4239 
Epoch 10/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8511 - loss: 0.4097 
Epoch 11/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8489 - loss: 0.4159 
Epoch 12/30
30/30 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy